### PR TITLE
New: `no-return-await` rule. (fixes #7537)

### LINKED ISSUE #7537: Rule proposal: `no-return-await`
**Please describe what the rule should do:**
Inside an `async function`, `return await` is useless. Since the return value of an `async function` is always wrapped in `Promise.resolve`, `return await` doesn't actually do anything except add an extra tick before the overarching promise resolves/rejects. This pattern is almost certainly due to programmer ignorance of the return semantics of `async function`s.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
async function foo() {
  return await bar;
}
```

in favor of:
```js
async function foo() {
  return bar;
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
People are very anxious to use `await`, and will likely overuse it. This rule prevents a common hazard, and also helps educate developers about how `async function` works.

### PR DESCRIPTION
**Pull request by ljharb**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[x] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**Please describe what the rule should do:**
Inside an `async function`, `return await` is useless. Since the return value of an `async function` is always wrapped in `Promise.resolve`, `return await` doesn't actually do anything except add an extra tick before the overarching promise resolves/rejects. This pattern is almost certainly due to programmer ignorance of the return semantics of `async function`s.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
async function foo() {
  return await bar;
}
```

in favor of:
```js
async function foo() {
  return bar;
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
People are very anxious to use `await`, and will likely overuse it. This rule prevents a common hazard, and also helps educate developers about how `async function` works.

**What changes did you make? (Give an overview)**
I added a rule file, a test file, and a docs file. The rule checks for return statements whose argument is an await expression.

**Is there anything you'd like reviewers to focus on?**
¯\\\_(ツ)_/¯

### CODE REVIEW

File: lib/rules/no-return-await.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                    context.report({
+                        node,
+                        loc: node.argument.loc,
+                        message: "`return await` from an `async function` is redundant - just `return` the value directly.",
+                    });
+                }
+            },
+
+        };
+    }
```
Comment (not-an-aardvark) [inline lib/rules/no-return-await.js]: I think this rule should check async arrow functions as well:

```js
async () => await foo;
```

File: tests/lib/rules/no-return-await.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+const rule = require("../../../lib/rules/no-return-await");
+const RuleTester = require("../../../lib/testers/rule-tester");
+
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+const parserOptions = { ecmaVersion: 2017 };
```
Comment (not-an-aardvark) [inline tests/lib/rules/no-return-await.js]: Nit: I think you can pass this option as a default to the `RuleTester` constructor, to avoid repeating it for every test.

```js
const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 2017 } });
```

File: lib/rules/no-return-await.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        schema: [
+        ]
+    },
+
+    create(context) {
+        return {
+            ReturnStatement(node) {
+                if (node.argument && node.argument.type === "AwaitExpression") {
+                    context.report({
+                        node,
```
Comment (not-an-aardvark) [inline lib/rules/no-return-await.js]: Nit: Maybe it would be better to report `node.argument` instead of `node`, (or even the `await` keyword with `sourceCode.getFirstToken(node.argument)` to make it clear that `await` is the problem.

Comment (ljharb) [inline lib/rules/no-return-await.js]: ah, good call. i'll do that.

File: tests/lib/rules/no-return-await.js (hunk 2 above)
Comment (ljharb) [inline tests/lib/rules/no-return-await.js]: ah nice, i'll do that thanks :-)

File: lib/rules/no-return-await.js (hunk 3 above)
Comment (ljharb) [inline lib/rules/no-return-await.js]: Turns out if i pass `node.argument` i get a type of AwaitExpression, but if i getFirstToken i get Identifier - I think `node.argument` is the right value here? unless you mean to use `getFirstToken(node)` instead of `node.argument`?

Comment (not-an-aardvark) [inline lib/rules/no-return-await.js]: No, `getFirstToken(node)` would get the `return` keyword rather than the `await` keyword.

After trying it out locally, it looks like acorn/espree is incorrectly identifying `await` as an identifier instead of a keyword:

```
Token {
  type: 'Identifier',
  value: 'await',
  start: 32,
  end: 37,
  loc:
   SourceLocation {
     start: Position { line: 3, column: 8 },
     end: Position { line: 3, column: 13 } },
  range: [ 32, 37 ] }
```

So `sourceCode.getFirstToken(node.argument)` is correct, but we should probably remove the type assertion from the tests for the time being.

Comment (ljharb) [inline lib/rules/no-return-await.js]: ah, so "AwaitExpression" is `await` + RHS, and getFirstToken would be just `await`?

Comment (not-an-aardvark) [inline lib/rules/no-return-await.js]: Right.

File: lib/rules/no-return-await.js (hunk 1 above)
Comment (ljharb) [inline lib/rules/no-return-await.js]: I'm having trouble figuring out how to get at an implicit return expression in an arrow function - any pointers?

File: tests/lib/rules/no-return-await.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            errors: [{
+                message: "`return await` from an async arrow function is redundant - just `return` the value directly.",
+                type: "Identifier", // pending https://github.com/eslint/espree/issues/304, should be "Keyword"
+            }],
+        },
+        {
+            code: "\nasync () => await bar()\n",
+            errors: [{
+                message: "await in an implicit-return async arrow function is redundant - just return the value directly.",
+                type: "Identifier", // pending https://github.com/eslint/espree/issues/304, should be "Keyword"
```
Comment (ljharb) [inline tests/lib/rules/no-return-await.js]: i left in these assertions, with a comment, so that when espree fixes its bug - these tests will fail, and they'll need to be updated to the correct values.

File: lib/rules/no-return-await.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    create(context) {
+        return {
+            ArrowFunctionExpression(node) {
+                if (node.async && node.body.type === "AwaitExpression") {
+                    const sourceCode = context.getSourceCode();
+
+                    context.report({
+                        node: sourceCode.getFirstToken(node.body),
+                        loc: node.body.loc,
+                        message: "await in an implicit-return async arrow function is redundant - just return the value directly.",
```
Comment (not-an-aardvark) [inline lib/rules/no-return-await.js]: Nit: This message (and the other one too) seems a bit wordy to me. What do you think about replacing it with something like `` Redundant use of `await` on a return value.``? I'm not sure that mentioning "implicit-return async arrow function" is helpful in telling users about the error.

File: lib/rules/no-return-await.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                        context.report({
+                            node: sourceCode.getFirstToken(argument),
+                            loc,
+                            message: "`return await` from an `async function` is redundant - just `return` the value directly.",
+                        });
+                    } else if (containingFunction.body.type === "BlockStatement") {
+                        context.report({
+                            node: sourceCode.getFirstToken(argument),
+                            loc,
+                            message: "`return await` from an async arrow function is redundant - just `return` the value directly.",
```
Comment (not-an-aardvark) [inline lib/rules/no-return-await.js]: Is there a reason for these messages to be different? While one of the functions is an arrow function, I'm not sure it's worth mentioning in the error message, since the fact that it's an arrow function isn't relevant to why the `await` statement is redundant.

File: lib/rules/no-return-await.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                    });
+                }
+            },
+            ReturnStatement(node) {
+                const argument = node.argument;
+
+                if (argument && argument.type === "AwaitExpression") {
+                    const sourceCode = context.getSourceCode();
+                    const loc = argument.loc;
+                    const containingFunction = node.parent.parent;
```
Comment (kaicataldo) [inline lib/rules/no-return-await.js]: Does this work for deeper nested await expressions? For example, cases like:

```js
async function foo() {
  if (a) {
    if (b) {
      return await bar;
    }
  }
}
```

Comment (ljharb) [inline lib/rules/no-return-await.js]: hm, that's a good point. i'll add that as a test case.

File: lib/rules/no-return-await.js (hunk 5 above)
Comment (ljharb) [inline lib/rules/no-return-await.js]: I guess it seems to me that error messages should be as specific as possible - I could certainly make them all the same, but it seems like a missed opportunity to avoid them having to look up the documentation.

Comment (not-an-aardvark) [inline lib/rules/no-return-await.js]: I guess just I don't understand how this particular piece of information is relevant. For example, one could theoretically make the error message `` Redundant use of `await` on a return value of 5 in a function called `foo` in another function called `bar` in a file called `baz.js` ``, but that would obviously be unhelpful; most of the information there is irrelevant, and it might mislead the user about what the rule is actually checking for.

The only thing that this rule *is* checking for is the use of `await` on return values; pointing out other features of the code, such as whether the statement is inside an arrow function, seems like it would draw the user's attention away from the main issue.

File: lib/rules/no-return-await.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            },
+            ReturnStatement(node) {
+                const argument = node.argument;
+
+                if (argument && argument.type === "AwaitExpression") {
+                    const sourceCode = context.getSourceCode();
+                    const loc = argument.loc;
+                    const containingFunction = getContainingFunction(node);
+
+                    if (containingFunction.type !== "ArrowFunctionExpression" || containingFunction.body.type === "BlockStatement") {
```
Comment (not-an-aardvark) [inline lib/rules/no-return-await.js]: I think this check can be removed; it applies to every return statement (since arrow functions with implicit return don't actually contain a `ReturnStatement`).

File: lib/rules/no-return-await.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    "ArrowFunctionExpression",
+    "MethodDefinition",
+]);
+
+/**
+ * @param {ASTNode} node the node from which to locate a containing function
+ * @returns {ASTNode} the containing function node
+ */
+function getContainingFunction(node) {
+    if (!node.parent || functionTypes.has(node.type)) {
```
Comment (not-an-aardvark) [inline lib/rules/no-return-await.js]: I don't think this function is necessary anymore (see my comment below), but if we do end up keeping it, it might be better to use [`astUtils.isFunction`](https://github.com/eslint/eslint/blob/e117b80cf2ed7b712057d3c2194c1c626edd3ce3/lib/ast-utils.js#L611-L613) to check whether a node is a function.

File: lib/rules/no-return-await.js (hunk 7 above)
Comment (ljharb) [inline lib/rules/no-return-await.js]: Fixed - i wrote a little abstraction to cleanly get the containing function that should handle all cases.

File: lib/rules/no-return-await.js (hunk 5 above)
Comment (ljharb) [inline lib/rules/no-return-await.js]: I guess I'll just make them the same and use your suggestion.

**Comment by kaicataldo** [comment]
Waiting to merge pending the conclusion of [this discussion](https://github.com/eslint/eslint/issues/6820#issuecomment-258761210)

**Comment by nzakas** [comment]
I think we're good to go on this. Thanks @ljharb!
